pkg/certwatcher/certwatcher_test.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certwatcher_test
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"net"
+	"os"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/certwatcher"
+)
+
+var _ = Describe("CertWatcher", func() {
+	var _ = Describe("certwatcher New", func() {
+		It("should errors without cert/key", func() {
+			_, err := certwatcher.New("", "")
+			Expect(err).ToNot(BeNil())
+		})
+	})
+
+	var _ = Describe("certwatcher Start", func() {
+		var (
+			ctx       context.Context
+			ctxCancel context.CancelFunc
+			watcher   *certwatcher.CertWatcher
+		)
+
+		BeforeEach(func() {
+			ctx, ctxCancel = context.WithCancel(context.Background())
+
+			err := writeCerts(certPath, keyPath, "127.0.0.1")
+			Expect(err).To(BeNil())
+
+			Eventually(func() error {
+				for _, file := range []string{certPath, keyPath} {
+					_, err := os.ReadFile(file)
+					if err != nil {
+						return err
+					}
+					continue
+				}
+
+				return nil
+			}).Should(Succeed())
+
+			watcher, _ = certwatcher.New(certPath, keyPath)
+		})
+
+		startWatcher := func() (done <-chan struct{}) {
+			doneCh := make(chan struct{})
+			go func() {
+				defer GinkgoRecover()
+				defer close(doneCh)
+				Expect(watcher.Start(ctx)).To(Succeed())
+			}()
+			// wait till we read first cert
+			Eventually(func() error {
+				err := watcher.ReadCertificate()
+				return err
+			}).Should(Succeed())
+			return doneCh
+		}
+
+		It("should read the initial cert/key", func() {
+			doneCh := startWatcher()
+
+			ctxCancel()
+			Eventually(doneCh, "4s").Should(BeClosed())
+		})
+
+		It("should reload currentCert when changed", func() {
+			doneCh := startWatcher()
+
+			firstcert, _ := watcher.GetCertificate(nil)
+
+			err := writeCerts(certPath, keyPath, "192.168.0.1")
+			Expect(err).To(BeNil())
+
+			Eventually(func() bool {
+				secondcert, _ := watcher.GetCertificate(nil)
+				first := firstcert.PrivateKey.(*rsa.PrivateKey)
+				return first.Equal(secondcert.PrivateKey)
+			}).ShouldNot(BeTrue())
+
+			ctxCancel()
+			Eventually(doneCh, "4s").Should(BeClosed())
+		})
+	})
+})
+
+func publicKey(priv interface{}) interface{} {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
```

Comment:
can we properly type the function signature if this effectively has to be a `*rsa.PrivateKey`?

Reply:
Yeah, I'll be honest cert gen was a bit of copy/pasta where they had the ability to create multiple algo based certs ecdsa/rsa etc and I just trimmed this down. Given this is only one algo I can just hard code this type on line https://github.com/kubernetes-sigs/controller-runtime/pull/1441/files/efd29024698fb095a8c004dea2641925655fb4ad#diff-cc4e056d68e41389001be7de46dfac27f508da1039369db677584b5721321ee4R162

```
derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, publicKey(priv), priv)
```

Reply:
Updated.